Runtime openGL loading

Hazardu opened the pull request.

Hazardu committed:
created files

Hazardu committed:
created files

Hazardu committed:
Prepared classes

Hazardu committed:
Prepared classes

Hazardu committed:
Merge branch 'runtimeOpenGLLoading' of https://github.com/Hazardu/IrrlichtBAW into runtimeOpenGLLoading

AnastaZIuk commented on source/Irrlicht/COpenGLFunctionTable.h:
```diff
@@ -0,0 +1,4530 @@
+			bool TextureCompressionExtension;
+
+			// Some non-boolean properties
+			static int32_t reqUBOAlignment;
```
not sure if the class should store anything associated with opengl state's properties

devshgraphicsprogramming replied:
Function table only stores function pointers!

AnastaZIuk replied:
yeah

devshgraphicsprogramming commented on source/Irrlicht/COpenGLFunctionTable.cpp:
```diff
@@ -0,0 +1,1500 @@
+		int32_t COpenGLFunctionTable::reqUBOAlignment = 0;
+		int32_t COpenGLFunctionTable::reqSSBOAlignment = 0;
+		int32_t COpenGLFunctionTable::reqTBOAlignment = 0;
+
+		uint64_t COpenGLFunctionTable::maxUBOSize = 0;
+		uint64_t COpenGLFunctionTable::maxSSBOSize = 0;
+		uint64_t COpenGLFunctionTable::maxTBOSizeInTexels = 0;
+		uint64_t COpenGLFunctionTable::maxBufferSize = 0;
+
+		int32_t COpenGLFunctionTable::minMemoryMapAlignment = 0;
+
+		int32_t COpenGLFunctionTable::MaxComputeWGSize[3]{ 0, 0, 0 };
+
+		uint32_t COpenGLFunctionTable::MaxArrayTextureLayers = 2048;
+		uint8_t COpenGLFunctionTable::MaxTextureUnits = 96;
+		uint8_t COpenGLFunctionTable::MaxAnisotropy = 8;
+		uint8_t COpenGLFunctionTable::MaxUserClipPlanes = 8;
+		uint8_t COpenGLFunctionTable::MaxMultipleRenderTargets = 4;
+		uint32_t COpenGLFunctionTable::MaxIndices = 65535;
+		uint32_t COpenGLFunctionTable::MaxVertices = 0xffffffffu;
+		uint32_t COpenGLFunctionTable::MaxVertexStreams = 1;
+		uint32_t COpenGLFunctionTable::MaxXFormFeedbackComponents = 64;
+		uint32_t COpenGLFunctionTable::MaxGPUWaitTimeout = 0;
+		uint32_t COpenGLFunctionTable::InvocationSubGroupSize[2] = { 32,64 };
+		GLuint COpenGLFunctionTable::SPIR_VextensionsCount = 0u;
+		core::smart_refctd_dynamic_array<const GLubyte*> COpenGLFunctionTable::SPIR_Vextensions;
+		uint32_t COpenGLFunctionTable::MaxGeometryVerticesOut = 65535;
+		float COpenGLFunctionTable::MaxTextureLODBias = 0.f;
+
+		uint32_t COpenGLFunctionTable::maxUBOBindings = 0u;
+		uint32_t COpenGLFunctionTable::maxSSBOBindings = 0u;
+		uint32_t COpenGLFunctionTable::maxTextureBindings = 0u;
+		uint32_t COpenGLFunctionTable::maxTextureBindingsCompute = 0u;
+		uint32_t COpenGLFunctionTable::maxImageBindings = 0u;
+
+		//uint32_t COpenGLFunctionTable::MaxXFormFeedbackInterleavedAttributes = GL_MAX_TRANSFORM_FEEDBACK_INTERLEAVED_COMPONENTS;
+		//uint32_t COpenGLFunctionTable::MaxXFormFeedbackSeparateAttributes = GL_MAX_TRANSFORM_FEEDBACK_SEPARATE_COMPONENTS;
+
+		bool COpenGLFunctionTable::IsIntelGPU = false;
+		bool COpenGLFunctionTable::needsDSAFramebufferHack = true;
```
all this you probably don't need

devshgraphicsprogramming commented on source/Irrlicht/COpenGLFunctionTable.h:
```diff
@@ -0,0 +1,4530 @@
+				pGlBeginConditionalRender(id, mode);
+		}
+
+		inline void COpenGLFunctionTable::extGlEndConditionalRender()
+		{
+			if (pGlEndConditionalRender)
+				pGlEndConditionalRender();
+		}
```
all the functions that don't have special workarounds, i.e. check for `Version` to choose between different function calls to use, etc.

You can remove these functions and make the variables like `pGlEndConditionalRender` public, BEWARE, for stuff that needs the fallbacks and special code... DO NOT MAKE THE FUNCTION POINTERS PUBLIC!

devshgraphicsprogramming commented on source/Irrlicht/COpenGLFunctionTable.cpp:
```diff
@@ -0,0 +1,1500 @@
+			/**
+			pGl = () IRR_OGL_SDL_LOAD_FUNCTIONPTR("gl");
+			**/
+			pGlProvokingVertex = (PFNGLPROVOKINGVERTEXPROC)IRR_OGL_SDL_LOAD_FUNCTIONPTR("glProvokingVertex");
```
My DLL Loader and DLL table C-preprocessor macros and classes that I've shown you, will do this automatically for you!

Hazardu replied:
So, the cast to whatever is in the parentheses is no longer required?
Im not sure if the macro generates any function within the class to initialize those pointers, if so, can this be done like this?
```
using LibLoader = system::DefaultFuncPtrLoader;
IRR_SYSTEM_DECLARE_DYNAMIC_FUNCTION_CALLER_CLASS(OpenGLMultiTexturing, LibLoader,
			glActiveTexture,
			//...
			)
OpenGLMultiTexturing multitexturing;
```

and in some function where i initialize the pointers i would use the `SDL_GL_GetProcAddress` in this fashion
```
multitexturing.pglActiveTexture = SDL_GL_GetProcAddress("glActiveTexture");
```

devshgraphicsprogramming replied:
> So, the cast to whatever is in the parentheses is no longer required?

It will do all the function pointer loading in the constructor "automagically".

> Im not sure if the macro generates any function within the class to initialize those pointers, if so, can this be done like this?

You need to change this out `using LibLoader = system::DefaultFuncPtrLoader;` for something else that you derive from the same base, because the default uses Windows/Unix dll system symbol retrieval functions, and for OpenGL special functions need to be used (I recommend the `SDL_GL_GetProcAddress`). 

Then everything is done automatically in constructors, then you don't need to do anything extra except for default constructing `multitexturing`.

I advise you use the special MSVC compiler flag `/P`
https://docs.microsoft.com/en-us/cpp/build/reference/p-preprocess-to-a-file?view=vs-2019
to see what these macros expand to (especially the already working OpenCL and CUDA implementations)

AnastaZIuk commented:
You should use the macros probably and as far as I remember split some functions into classes creating them by macro as well. Look at glue directive

devshgraphicsprogramming commented:
> 
> 
> You should use the macros probably and as far as I remember split some functions into classes creating them by macro as well. Look at glue directive

As you may recall from our chat @Hazardu I gave examples of how its done in `CCudaHandler` and `COpenCLHandler`

Hazardu committed:
Update COpenGLFunctionTable.h

Hazardu committed:
separation into classes

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -552,1338 +553,579 @@ namespace irr
+			{
+			public:
+
+class OpenGLFunctionLoader final : system::FuncPtrLoader
```
just make sure the indendation is still nice ;)

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -552,1338 +553,579 @@ namespace irr
+	{
+		return pGlGetProcAdress(funcname);
+	}
+	IRR_SYSTEM_DECLARE_DYNAMIC_FUNCTION_CALLER_CLASS(GLsync, OpenGLFunctionLoader
+		,glFenceSync
+		,glDeleteSync,
+		,glClientWaitSync
+		,glWaitSync
+		);
+	IRR_SYSTEM_DECLARE_DYNAMIC_FUNCTION_CALLER_CLASS(GLbuffers, OpenGLFunctionLoader
```
these should not be members of `OpenGLFunctionLoader`, should be declared outside

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -552,1338 +553,579 @@ namespace irr
+		, glBindBufferRange
+		, glBindBuffersBase
+		, glBindBuffersRange
+		, glBlitNamedFramebuffer
+		, glBlitFramebuffer
+		, glDeleteFramebuffers
+		, glCreateFramebuffers
+		, glGenFramebuffers
+		, glBindFramebuffer
+		, glCheckFramebufferStatus
+		, glCheckNamedFramebufferStatus
+		, glCheckNamedFramebufferStatusEXT
+		, glFramebufferTexture
+		, glNamedFramebufferTexture
+		, glNamedFramebufferTextureEXT
+		, glFramebufferTextureLayer
+		, glNamedFramebufferTextureLayer
+		, glNamedFramebufferTextureLayerEXT
+		, glFramebufferTexture2D
+		, glNamedFramebufferTexture2DEXT
+		, glDrawBuffers
+		, glNamedFramebufferDrawBuffers
+		, glFramebufferDrawBuffersEXT
+		, glNamedFramebufferDrawBuffer
+		, glFramebufferDrawBufferEXT
+		, glNamedFramebufferReadBuffer
+		, glFramebufferReadBufferEXT
+		, glClearNamedFramebufferiv
+		, glClearNamedFramebufferuiv
+		, glClearNamedFramebufferfv
+		, glClearNamedFramebufferfi
+		, glClearBufferiv
+		, glClearBufferuiv
+		, glClearBufferfv
+		, glClearBufferfi
```
framebuffer != buffer

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -552,1338 +553,579 @@ namespace irr
+
+
+
+//GLOS
+wglSwapIntervalEXT
+glXSwapIntervalSGI
+glXSwapIntervalEXT
+glXSwapIntervalMESA
```
you will need to declare these separate under `ifdef` that check for currently compiling system

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -552,1338 +553,579 @@ namespace irr
+
+class OpenGLFunctionLoader final : system::FuncPtrLoader
+{
+	using FUNC_PTR_TYPE = decltype(SDL_GL_GetProcAddress)*;
+	FUNC_PTR_TYPE pGlGetProcAdress;//pClGetExtensionFunctionAddress;
```
you can add and statically link SDL2 to the engine now via CMake, I already added the submodule

Hazardu committed:
Update COpenGLFunctionTable.h

Hazardu committed:
Reordered function names to make more sense

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -40,3733 +15,3084 @@ wywali
-				glBindTexture(target, texture);
-				pGlGenerateMipmap(target);
-				glBindTexture(target, bound);
+				//std::swap(pGlGetProcAdress, other.pGlGetProcAdress);
```
you can swap entire declared pointer classes

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -40,3733 +15,3084 @@ wywali
+			,glProvokingVertex
+		);
+		IRR_SYSTEM_DECLARE_DYNAMIC_FUNCTION_CALLER_CLASS(GLdrawing, OpenGLFunctionLoader
+			,glPrimitiveRestartIndex
```
this sets a pipeline state, put it together with the shaders

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -40,3733 +15,3084 @@ wywali
+#endif
+		);
+		IRR_SYSTEM_DECLARE_DYNAMIC_FUNCTION_CALLER_CLASS(GLprogram, OpenGLFunctionLoader
+			,glEnablei
+			,glDisablei
```
you also need to get entry points for "vanilla" OpenGL 1.4 functions that we use, such as `glEnable`.... oh and btw `Enable/Disable` are common opengl functions they handle everything

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -40,3733 +15,3084 @@ wywali
+			,glDisablei
+			,glProgramParameteri
+			,glPatchParameterfv
+			,glPatchParameteri
```
these are pipeline things, so should go together with the shader stuff

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -40,3733 +15,3084 @@ wywali
-			return 0ull;
+			if(SDL_GL_LoadLibrary(nullptr);) //nullptr to load the default OpenGL library
+				//log error
+			if(SDL_VideoInit())
```
its not the function table that should do it, it should be the library loader instead

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -40,3733 +15,3084 @@ wywali
+	}
+	//
+	//#pragma region Inline Implementations
+	//		//TODO use khronos.org to add params & add ptr null checks
```
you dont need to check for nullptr anymore

Hazardu committed:
Dont have much time today, doing home renovations, but i will try to finish this asap

Hazardu committed:
Update COpenGLFunctionTable.h

I Couldnt find defintions for the following fuctions online:
GlFramebufferDrawBuffersEXT
GlFramebufferDrawBufferEXT
GlFramebufferReadBufferEXT
GlGetNamedBufferSubDataEXT

The pace of writing those functions is quite slow, as im checking what each of them does

There also have been some on the wiki in related articles that i have added, for ex: glGet functions had a few more types on the wiki page

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -19,3116 +19,5702 @@
+			/// </summary>
+			inline void GlGetBooleani_v(GLenum pname, GLuint index, GLboolean* data)
+			{
+				glGeneral.glGetBooleani_v(pname, index, data);
+			}
+			/// <summary>
+			/// return the value or values of a selected parameter
+			/// </summary>
+			inline void GlGetBooleanv(GLenum pname, GLboolean* data)
+			{
+				glGeneral.glGetBooleanv(pname, data);
+			}
+			/// <summary>
+			/// return the value or values of a selected parameter
+			/// </summary>
+			inline void GlGetFloati_v(GLenum pname, GLuint index, float* data)
+			{
+				glGeneral.pglGetFloati_v(pname, index, data);
+			}
+			/// <summary>
+			/// return the value or values of a selected parameter
+			/// </summary>
+			inline void GlGetFloatv(GLenum pname, float* data)
+			{
+				glGeneral.pglGetFloatv(pname, data);
+			}
+			/// <summary>
+			/// return the value or values of a selected parameter
+			/// </summary>
+			inline void GlGetInteger64v(GLenum pname, GLint64* data)
+			{
+				glGeneral.pglGetInteger64v(pname, data);
+			}
+			/// <summary>
+			/// return the value or values of a selected parameter
+			/// </summary>
+			inline void GlGetIntegeri_v(GLenum pname, GLuint index, GLint* data)
+			{
+				glGeneral.pglGetIntegeri_v(pname, index, data);
+			}
```
This is code bloat.

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -19,3116 +19,5702 @@
+			/// <summary>
+			/// 
+			/// </summary>
+			inline void GlActiveTexture()
+			{
+				glTexture.pglActiveTexture();
+			}
+			/// <summary>
+			/// 
+			/// </summary>
+			inline void GlBindTextures()
+			{
+				glTexture.pglBindTextures();
+			}
```
I don't think I want these public, nor their pointers

Hazardu replied:
Should i use the macro to create another class with pointers again but this time keep the class private?

Hazardu committed:
removed one line functions

Hazardu committed:
added functions that need to be emulated

todo: move the ptrs used in those classes to a private class

Hazardu committed:
In the middle of resolving issues related to glext.h

Does not compile yet, i need to remove include glext.h from header files, then resolve issue of unexisting type declarations like GLint

Hazardu committed:
Removed Extension handler, put functions into function table and fields to COpenGLDriver.h

Does not compile yet, inside glxext.h there are uses of types like Display, which arent defined. Similiar problem was occuring in glext, but was solved by including <GL/gl.h>.
There doesnt appear to be a file like <GL/glx.h>

Hazardu committed:
Replaced uses of COpenGLExtensionHandler.h with COpenGlFunctionTable.h

Hazardu committed:
Progress on Function table, replaced uses of deprecated Extensionhandler in COpenGLDRiver, partly in other files

Theres a few errors in CopenGLDriver.cpp
first being at line 233, Function Table is not a nonstatic data member or base class of class
next, 1072, initExtensions() no longer exist, so i assume i need to rewrite it, and put it inside COpenGLFunctionTable.h
3rd issue is RUNNING_IN_RENDERDOC_EXTENSION_NAME at line 1181 is undefined
lastly deltype at 2395 with `&COpenGLFunctionTable::glGeneral.pglDisablei`

Hazardu commented:
in ExtensionHandler, there is a method initExtensions thats being invoked from COpenGLDriver.genericDriverInit. Should i rewrite it and put it inside FunctionTable or OpenGLDriver?

Hazardu committed:
Added initExtensions function

devshgraphicsprogramming commented on include/COpenGLStateManagerImpl.h:
```diff
@@ -61,8 +61,8 @@
 #define glBindBufferRange_MACRO glBindBufferRange
 #endif // glBindBufferRange_MACRO
 
-#ifndef glGetNamedBufferParameteri64v_MACRO
-#define glGetNamedBufferParameteri64v_MACRO glGetNamedBufferParameteri64v
+#ifndef glGetNamedBufferParameteri64v_MACRO
+#define glGetNamedBufferParameteri64v_MACRO glGetNamedBufferParameteri64v
 #endif
```
all these we don't need, actually I don't think we need the `COpenGLStateManagerImpl` and the `COpenGLStateManager` at all

devshgraphicsprogramming commented on source/Irrlicht/COpenGLFrameBuffer.cpp:
```diff
@@ -180,7 +180,7 @@ bool COpenGLFrameBuffer::attach(E_FBO_ATTACHMENT_POINT attachmenPoint, core::sma
     if (enabledBufferCnt)
         enabledBufferCnt += 1-EFAP_COLOR_ATTACHMENT0;
-    COpenGLExtensionHandler::extGlNamedFramebufferDrawBuffers(frameBuffer, enabledBufferCnt, drawBuffers);
+    COpenGLFunctionTable::extGlNamedFramebufferDrawBuffers(frameBuffer, enabledBufferCnt, drawBuffers);
```
nope the methods of the function table cannot be static, you need `getThreadContext_helper` then get the function table from the `SAuxContext` and then call the non-static method

devshgraphicsprogramming replied:
confer with https://github.com/buildaworldnet/IrrlichtBAW/blob/6b7aeccf2d8030799992b651fcd5546d10f79e48/source/Irrlicht/COpenGLDriver.h#L586

devshgraphicsprogramming commented on source/Irrlicht/COpenGLExtensionHandler.cpp:
```diff
@@ -1,1564 +1,1564 @@
+//}
+//}
+//
+//#endif
```
just remove the file from the repo, we have a file history after all

devshgraphicsprogramming commented on src/irr/video/IOpenGLPipeline.h:
```diff
@@ -25,8 +25,8 @@ class IOpenGLPipeline
                         continue;
-                    const GLuint GLname = COpenGLExtensionHandler::extGlCreateProgram();
-                    COpenGLExtensionHandler::extGlProgramBinary(GLname, bin.format, bin.binary->data(), static_cast<uint32_t>(bin.binary->size()));
+                    const GLuint GLname = COpenGLFunctionTable::glShader.pglCreateProgram();
+                    COpenGLFunctionTable::glShader.pglProgramBinary(GLname, bin.format, bin.binary->data(), static_cast<uint32_t>(bin.binary->size()));
```
no static methods

devshgraphicsprogramming commented on src/irr/video/IOpenGLPipeline.h:
```diff
@@ -38,7 +38,7 @@ class IOpenGLPipeline
             for (const auto& p : (*m_GLprograms))
                 if (p.GLname != 0u)
-                    COpenGLExtensionHandler::extGlDeleteProgram(p.GLname);
+                    COpenGLFunctionTable::extGlDeleteProgram(p.GLname);
```
no static methods

devshgraphicsprogramming commented on src/irr/video/IOpenGLPipeline.h:
```diff
@@ -54,7 +54,7 @@ class IOpenGLPipeline
             IRR_ASSUME_ALIGNED(_pcData, 128);
 
-            using gl = COpenGLExtensionHandler;
+            using gl = COpenGLFunctionTable;
```
no static methods

devshgraphicsprogramming commented on src/irr/video/COpenGLSpecializedShader.cpp:
```diff
@@ -173,10 +173,10 @@ auto COpenGLSpecializedShader::compile(const COpenGLPipelineLayout* _layout, con
 
-	GLuint GLname = COpenGLExtensionHandler::extGlCreateShaderProgramv(m_GLstage, 1u, &glslCode_cstr);
+	GLuint GLname = COpenGLFunctionTable::extGlCreateShaderProgramv(m_GLstage, 1u, &glslCode_cstr);
 
 	GLchar logbuf[1u<<12]; //4k
-	COpenGLExtensionHandler::extGlGetProgramInfoLog(GLname, sizeof(logbuf), nullptr, logbuf);
+	COpenGLFunctionTable::extGlGetProgramInfoLog(GLname, sizeof(logbuf), nullptr, logbuf);
```
no static methods

devshgraphicsprogramming commented on src/irr/video/COpenGLSpecializedShader.cpp:
```diff
@@ -185,9 +185,9 @@ auto COpenGLSpecializedShader::compile(const COpenGLPipelineLayout* _layout, con
 	GLint binaryLength = 0;
-	COpenGLExtensionHandler::extGlGetProgramiv(GLname, GL_PROGRAM_BINARY_LENGTH, &binaryLength);
+	COpenGLFunctionTable::extGlGetProgramiv(GLname, GL_PROGRAM_BINARY_LENGTH, &binaryLength);
 	binary.binary = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<uint8_t>>(binaryLength);
-	COpenGLExtensionHandler::extGlGetProgramBinary(GLname, binaryLength, nullptr, &binary.format, binary.binary->data());
+	COpenGLFunctionTable::extGlGetProgramBinary(GLname, binaryLength, nullptr, &binary.format, binary.binary->data());
```
no static methods

devshgraphicsprogramming commented on src/irr/video/COpenGLSpecializedShader.cpp:
```diff
@@ -196,7 +196,7 @@ void COpenGLSpecializedShader::gatherUniformLocations(GLuint _GLname) const
 	m_locations.resize(m_uniformsList.size());
 	for (size_t i = 0ull; i < m_uniformsList.size(); ++i)
-		m_locations[i] = COpenGLExtensionHandler::extGlGetUniformLocation(_GLname, m_uniformsList[i].m.name.c_str());
+		m_locations[i] = COpenGLFunctionTable::extGlGetUniformLocation(_GLname, m_uniformsList[i].m.name.c_str());
```
no static methods

devshgraphicsprogramming commented on src/irr/video/COpenGLPipelineCache.cpp:
```diff
@@ -82,12 +82,12 @@ core::smart_refctd_ptr<asset::ICPUPipelineCache> COpenGLPipelineCache::convertTo
-		uuid += std::to_string(COpenGLExtensionHandler::Version);
+		uuid += std::to_string(COpenGLFunctionTable::Version);
 
 		std::string exts;
-		for (uint32_t k = 0u; k < COpenGLExtensionHandler::IRR_OpenGL_Feature_Count; ++k)
+		for (uint32_t k = 0u; k < COpenGLFunctionTable::IRR_OpenGL_Feature_Count; ++k)
 		{
-			if (COpenGLExtensionHandler::FeatureAvailable[k])
+			if (COpenGLFunctionTable::FeatureAvailable[k])
```
these need to be fields of the `COpenGLDriver` actually, not the table as the table represents variables about each individual GL context, we cannot have divergent feature support across threads.

devshgraphicsprogramming commented on src/irr/video/COpenGLImageView.h:
```diff
@@ -49,15 +49,15 @@ class COpenGLImageView final : public IGPUImageView
 			}
-			COpenGLExtensionHandler::extGlTextureParameterIuiv(name,target,GL_TEXTURE_SWIZZLE_RGBA,swizzle);
+			COpenGLFunctionTable::extGlTextureParameterIuiv(name,target,GL_TEXTURE_SWIZZLE_RGBA,swizzle);
 		}
 
 		void regenerateMipMapLevels() override
 		{
 			if (params.subresourceRange.levelCount <= 1u)
 				return;
 
-			COpenGLExtensionHandler::extGlGenerateTextureMipmap(name,target);
+			COpenGLFunctionTable::extGlGenerateTextureMipmap(name,target);
```
no static methods

devshgraphicsprogramming commented on src/irr/video/COpenGLImageView.h:
```diff
@@ -35,9 +35,9 @@ class COpenGLImageView final : public IGPUImageView
             assert(internalFormat != GL_INVALID_ENUM);
-			//COpenGLExtensionHandler::extGlCreateTextures(target, 1, &name);
+			//COpenGLFunctionTable::extGlCreateTextures(target, 1, &name);
 			glGenTextures(1, &name);
-			COpenGLExtensionHandler::extGlTextureView(	name, target, static_cast<COpenGLImage*>(params.image.get())->getOpenGLName(), internalFormat, 
+			COpenGLFunctionTable::extGlTextureView(	name, target, static_cast<COpenGLImage*>(params.image.get())->getOpenGLName(), internalFormat, 
```
no static methods

devshgraphicsprogramming commented on src/irr/video/COpenGLImage.h:
```diff
@@ -37,27 +37,27 @@ class COpenGLImage final : public IGPUImage, public IDriverMemoryAllocation
+					COpenGLFunctionTable::extGlTextureStorage2D(	name, target, params.mipLevels, internalFormat,
 																	params.extent.width, params.arrayLayers);
 					break;
 				case IGPUImage::ET_2D:
 					target = GL_TEXTURE_2D_ARRAY;
-					COpenGLExtensionHandler::extGlCreateTextures(target, 1, &name);
-					COpenGLExtensionHandler::extGlTextureStorage3D(	name, target, params.mipLevels, internalFormat,
+					COpenGLFunctionTable::extGlCreateTextures(target, 1, &name);
+					COpenGLFunctionTable::extGlTextureStorage3D(	name, target, params.mipLevels, internalFormat,
 																	params.extent.width, params.extent.height, params.arrayLayers);
 					break;
 				case IGPUImage::ET_3D:
 					target = GL_TEXTURE_3D;
-					COpenGLExtensionHandler::extGlCreateTextures(target, 1, &name);
-					COpenGLExtensionHandler::extGlTextureStorage3D(	name, target, params.mipLevels, internalFormat,
+					COpenGLFunctionTable::extGlCreateTextures(target, 1, &name);
+					COpenGLFunctionTable::extGlTextureStorage3D(	name, target, params.mipLevels, internalFormat,
```
no static methods

devshgraphicsprogramming commented on src/irr/video/COpenGLImage.h:
```diff
@@ -37,27 +37,27 @@ class COpenGLImage final : public IGPUImage, public IDriverMemoryAllocation
 		{
 			#ifdef OPENGL_LEAK_DEBUG
-				COpenGLExtensionHandler::textureLeaker.registerObj(this);
+				COpenGLFunctionTable::textureLeaker.registerObj(this);
```
you can get rid of the `textureLeaker` we will replace it with something better in the future

devshgraphicsprogramming commented on src/irr/video/COpenGLBufferView.h:
```diff
@@ -19,13 +19,13 @@ class COpenGLBufferView : public IGPUBufferView
-			COpenGLExtensionHandler::extGlCreateTextures(GL_TEXTURE_BUFFER, 1, &m_textureName);
+			COpenGLFunctionTable::extGlCreateTextures(GL_TEXTURE_BUFFER, 1, &m_textureName);
 			m_GLformat = getSizedOpenGLFormatFromOurFormat(m_format);
 
 			if (m_offset==0u && m_size==m_buffer->getSize())
-				COpenGLExtensionHandler::extGlTextureBuffer(m_textureName, m_GLformat, static_cast<COpenGLBuffer*>(m_buffer.get())->getOpenGLName());
+				COpenGLFunctionTable::extGlTextureBuffer(m_textureName, m_GLformat, static_cast<COpenGLBuffer*>(m_buffer.get())->getOpenGLName());
 			else
-				COpenGLExtensionHandler::extGlTextureBufferRange(m_textureName, m_GLformat, static_cast<COpenGLBuffer*>(m_buffer.get())->getOpenGLName(), m_offset, m_size);
+				COpenGLFunctionTable::extGlTextureBufferRange(m_textureName, m_GLformat, static_cast<COpenGLBuffer*>(m_buffer.get())->getOpenGLName(), m_offset, m_size);
```
no static methods

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -230,14 +230,22 @@ namespace video
 COpenGLDriver::COpenGLDriver(const irr::SIrrlichtCreationParameters& params,
 		io::IFileSystem* io, CIrrDeviceWin32* device, const asset::IGLSLCompiler* glslcomp)
-: CNullDriver(device, io, params), COpenGLExtensionHandler(),
+: CNullDriver(device, io, params), //COpenGLFunctionTable(),   // What to do here?
```
`COpenGLFunctionTable` shall belong to the `SAuxContext`

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -624,7 +632,7 @@ bool COpenGLDriver::initDriver(CIrrDeviceWin32* device)
 	genericDriverInit(device->getAssetManager());
 
-	extGlSwapInterval(Params.Vsync ? 1 : 0);
+	COpenGLFunctionTable::extGlSwapInterval(Params.Vsync ? 1 : 0);
```
no static methods

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -753,7 +769,7 @@ bool COpenGLDriver::initDriver(CIrrDeviceLinux* device, SAuxContext* auxCtxts)
 	// set vsync
 	//if (queryOpenGLFeature(IRR_))
-        extGlSwapInterval(Params.Vsync ? -1 : 0);
+        COpenGLFunctionTable::glShader.pglSwapInterval(Params.Vsync ? -1 : 0);
```
no static methods

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -942,7 +958,7 @@ void COpenGLDriver::cleanUpContextBeforeDelete()
 
     found->CurrentRendertargetSize = Params.WindowSize;
-    extGlBindFramebuffer(GL_FRAMEBUFFER, 0);
+    COpenGLFunctionTable::glFrameBuffer.pglBindFramebuffer(GL_FRAMEBUFFER, 0);
```
no static methods

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -951,17 +967,17 @@ void COpenGLDriver::cleanUpContextBeforeDelete()
     for (auto& vao : found->VAOMap)
     {
-        extGlDeleteVertexArrays(1, &vao.second.GLname);
+        COpenGLFunctionTable::glVertex.pglDeleteVertexArrays(1, &vao.second.GLname);
     }
     found->VAOMap.clear();
 
-    extGlUseProgram(0);
-    extGlBindProgramPipeline(0);
+    COpenGLFunctionTable::glShader.pglUseProgram(0);
+    COpenGLFunctionTable::glShader.pglBindProgramPipeline(0);
     for (auto& ppln : found->GraphicsPipelineMap)
-        extGlDeleteProgramPipelines(1, &ppln.second.GLname);
+        COpenGLFunctionTable::glShader.pglDeleteProgramPipelines(1, &ppln.second.GLname);
```
no static methods

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -1088,12 +1104,12 @@ bool COpenGLDriver::genericDriverInit(asset::IAssetManager* assMgr)
 	glViewport(0, 0, Params.WindowSize.Width, Params.WindowSize.Height);
 
 	// adjust flat coloring scheme to DirectX version
-	///extGlProvokingVertex(GL_FIRST_VERTEX_CONVENTION_EXT);
+	///COpenGLFunctionTable::glShader.pglProvokingVertex(GL_FIRST_VERTEX_CONVENTION_EXT);
 
 	// We need to reset once more at the beginning of the first rendering.
 	// This fixes problems with intermediate changes to the material during texture load.
     SAuxContext* found = getThreadContext_helper(false);
-	extGlClipControl(GL_UPPER_LEFT, GL_ZERO_TO_ONE); //once set and should never change (engine doesnt track it)
+    COpenGLFunctionTable::glGeneral.pglClipControl(GL_UPPER_LEFT, GL_ZERO_TO_ONE); //once set and should never change (engine doesnt track it)
     glEnable(GL_FRAMEBUFFER_SRGB);//once set and should never change (engine doesnt track it)
     glEnable(GL_TEXTURE_CUBE_MAP_SEAMLESS);//once set and should never change (engine doesnt track it)
     glDepthRange(1.0, 0.0);//once set and should never change (engine doesnt track it)
```
all this setup should probably move to context intialization, don't worry about it now we'll tackle it in a separate PR

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -1177,8 +1193,8 @@ const core::smart_refctd_dynamic_array<std::string> COpenGLDriver::getSupportedG
                 (*m_supportedGLSLExtsNames)[i++] = OpenGLFeatureStrings[m_GLSLExtensions[j]];
-        if (runningInRenderDoc)
-            (*m_supportedGLSLExtsNames)[i] = RUNNING_IN_RENDERDOC_EXTENSION_NAME;
+        //if (runningInRenderDoc)
+        //    (*m_supportedGLSLExtsNames)[i] = RUNNING_IN_RENDERDOC_EXTENSION_NAME;
```
I need this working before the merge

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -1248,7 +1264,7 @@ bool COpenGLDriver::dispatch(uint32_t _groupCountX, uint32_t _groupCountY, uint3
     ctx->flushStateCompute(GSB_PIPELINE | GSB_DESCRIPTOR_SETS | GSB_PUSH_CONSTANTS);
 
-    extGlDispatchCompute(_groupCountX, _groupCountY, _groupCountZ);
+    COpenGLFunctionTable::glCompute.pglDispatchCompute(_groupCountX, _groupCountY, _groupCountZ);
```
no static methods

`ctx->getGLFunctionTable().glCompute.`

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -3282,9 +3302,118 @@ void COpenGLDriver::enableClipPlane(uint32_t index, bool enable)
 
+void COpenGLDriver::initExtensions(bool stencilBuffer)
+{
+    core::stringc vendorString = (char*)glGetString(GL_VENDOR);
```
you need to initExtensions after you init the function table for the main context

devshgraphicsprogramming replied:
because `glGetString` needs to be called from the pointer in the function table

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -3282,9 +3302,118 @@ void COpenGLDriver::enableClipPlane(uint32_t index, bool enable)
+
+	GLint num = 0;
+
+	glGetIntegerv(GL_UNIFORM_BUFFER_OFFSET_ALIGNMENT, &reqUBOAlignment);
+    assert(core::is_alignment(reqUBOAlignment));
+	glGetIntegerv(GL_SHADER_STORAGE_BUFFER_OFFSET_ALIGNMENT, &reqSSBOAlignment);
+    assert(core::is_alignment(reqSSBOAlignment));
+	glGetIntegerv(GL_TEXTURE_BUFFER_OFFSET_ALIGNMENT, &reqTBOAlignment);
```
`glGetIntegerv` needs to be called from the pointer in the function table

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -3282,9 +3302,118 @@ void COpenGLDriver::enableClipPlane(uint32_t index, bool enable)
+    COpenGLFunctionTable::glGeneral.pglGetInteger64v(GL_MAX_TEXTURE_BUFFER_SIZE, reinterpret_cast<GLint64*>(&maxTBOSizeInTexels));
+    maxBufferSize = std::max(maxUBOSize, maxSSBOSize);
+
+    glGetIntegerv(GL_MAX_COMBINED_UNIFORM_BLOCKS, reinterpret_cast<GLint*>(&maxUBOBindings));
+    glGetIntegerv(GL_MAX_COMBINED_SHADER_STORAGE_BLOCKS, reinterpret_cast<GLint*>(&maxSSBOBindings));
+    glGetIntegerv(GL_MAX_COMBINED_TEXTURE_IMAGE_UNITS, reinterpret_cast<GLint*>(&maxTextureBindings));
+    glGetIntegerv(GL_MAX_COMPUTE_TEXTURE_IMAGE_UNITS, reinterpret_cast<GLint*>(&maxTextureBindingsCompute));
+    glGetIntegerv(GL_MAX_COMBINED_IMAGE_UNIFORMS, reinterpret_cast<GLint*>(&maxImageBindings));
+
+	glGetIntegerv(GL_MIN_MAP_BUFFER_ALIGNMENT, &minMemoryMapAlignment);
```
`glGetIntegerv` needs to be called from the pointer in the function table

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -3282,9 +3302,118 @@ void COpenGLDriver::enableClipPlane(uint32_t index, bool enable)
+	{
+		glGetIntegerv(GL_MAX_TEXTURE_MAX_ANISOTROPY_EXT, &num);
+		MaxAnisotropy = static_cast<uint8_t>(num);
+	}
+
+
+    if (FeatureAvailable[IRR_ARB_geometry_shader4])
+    {
+        glGetIntegerv(GL_MAX_GEOMETRY_OUTPUT_VERTICES, &num);
+        MaxGeometryVerticesOut = static_cast<uint32_t>(num);
+    }
+
+	if (FeatureAvailable[IRR_EXT_texture_lod_bias])
+		glGetFloatv(GL_MAX_TEXTURE_LOD_BIAS_EXT, &MaxTextureLODBias);
+
+
+	glGetIntegerv(GL_MAX_CLIP_DISTANCES, &num);
+	MaxUserClipPlanes=static_cast<uint8_t>(num);
+    glGetIntegerv(GL_MAX_DRAW_BUFFERS, &num);
+    MaxMultipleRenderTargets = static_cast<uint8_t>(num);
+
+	glGetFloatv(GL_ALIASED_LINE_WIDTH_RANGE, DimAliasedLine);
+	glGetFloatv(GL_ALIASED_POINT_SIZE_RANGE, DimAliasedPoint);
+	glGetFloatv(GL_SMOOTH_LINE_WIDTH_RANGE, DimSmoothedLine);
+	glGetFloatv(GL_SMOOTH_POINT_SIZE_RANGE, DimSmoothedPoint);
+
+    const GLubyte* shaderVersion = glGetString(GL_SHADING_LANGUAGE_VERSION);
+    float sl_ver;
+    sscanf(reinterpret_cast<const char*>(shaderVersion),"%f",&sl_ver);
+    ShaderLanguageVersion = static_cast<uint16_t>(core::round(sl_ver*100.0f));
+
+
+	//! For EXT-DSA testing
+	if (IsIntelGPU)
+	{
+		Version = 440;
+		FeatureAvailable[IRR_ARB_direct_state_access] = false;
+	}
+    num=0;
+    glGetIntegerv(GL_MAX_COMBINED_TEXTURE_IMAGE_UNITS, &num);
```
glGetIntegerv needs to be called from the pointer in the function table

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -3282,9 +3302,118 @@ void COpenGLDriver::enableClipPlane(uint32_t index, bool enable)
+	{
+		// undocumented flags, so use the RAW values
+		GLint val;
+		glGetIntegerv(0x9047, &val);
+		os::Printer::log("Dedicated video memory (kB)", std::to_string(val));
+		glGetIntegerv(0x9048, &val);
+		os::Printer::log("Total video memory (kB)", std::to_string(val));
+		glGetIntegerv(0x9049, &val);
+		os::Printer::log("Available video memory (kB)", std::to_string(val));
+	}
+	if (FeatureAvailable[IRR_ATI_meminfo])
+	{
+		GLint val[4];
+		glGetIntegerv(GL_TEXTURE_FREE_MEMORY_ATI, val);
+		os::Printer::log("Free texture memory (kB)", std::to_string(val[0]));
+		glGetIntegerv(GL_VBO_FREE_MEMORY_ATI, val);
+		os::Printer::log("Free VBO memory (kB)", std::to_string(val[0]));
+		glGetIntegerv(GL_RENDERBUFFER_FREE_MEMORY_ATI, val);
```
glGetIntegerv needs to be called from the pointer in the function table

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.h:
```diff
@@ -217,6 +722,609 @@ class COpenGLDriver final : public CNullDriver, public COpenGLExtensionHandler
 	public:
         struct SAuxContext;
 
+		enum EOpenGLFeatures {
+			IRR_3DFX_multisample = 0,
+			IRR_3DFX_tbuffer,
+			IRR_3DFX_texture_compression_FXT1,
```
this is on the border of making our code a little illegible, maybe create a class `src/irr/video/COpenGLFeatureMap` and shove all this extension stuff there and then make `COpenGLDriver` inherit from it?

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.h:
```diff
@@ -217,6 +722,609 @@ class COpenGLDriver final : public CNullDriver, public COpenGLExtensionHandler
+				
+		static bool FeatureAvailable[IRR_OpenGL_Feature_Count];
+
+		static bool IsIntelGPU;
+		static bool needsDSAFramebufferHack;
```
these two should be in the function table, since these variables control the behaviour of the `extGl` workarounds

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -211,7 +228,7 @@ namespace irr {
 				, glGetTextureHandleARB
 				, glGetTextureSamplerHandleARB
-				, glMakeTextureHandleResidentAR
+				//, glMakeTextureHandleResidentAR
```
ARB

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -220,7 +237,7 @@ namespace irr {
 				, glGetTextureHandleNV
 				, glGetTextureSamplerHandleNV
-				, glMakeTextureHandleResidentAR
+				//, glMakeTextureHandleResidentAR
```
ARB

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -1,27 +1,45 @@
+//#include "../../3rdparty/GL/glext.h"
 
+//#ifdef __gl_glext_h_
+//#error chuj
```
this is going in the public git repo history forever ;)

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -464,87 +484,87 @@ namespace irr {
+			static void extGlTextureStorage1D(GLuint texture, GLenum target, GLsizei levels, GLenum internalformat, GLsizei width);
+			static void extGlTextureStorage2D(GLuint texture, GLenum target, GLsizei levels, GLenum internalformat, GLsizei width, GLsizei height);
+			static void extGlTextureStorage3D(GLuint texture, GLenum target, GLsizei levels, GLenum internalformat, GLsizei width, GLsizei height, GLsizei depth);
+			static void extGlTextureStorage2DMultisample(GLuint texture, GLenum target, GLsizei samples, GLenum internalformat, GLsizei width, GLsizei height, GLboolean fixedsamplelocations);
+			static void extGlTextureStorage3DMultisample(GLuint texture, GLenum target, GLsizei samples, GLenum internalformat, GLsizei width, GLsizei height, GLsizei depth, GLboolean fixedsamplelocations);
+			static void extGlGetTextureSubImage(GLuint texture, GLint level, GLint xoffset, GLint yoffset, GLint zoffset, GLsizei width, GLsizei height, GLsizei depth, GLenum format, GLenum type, GLsizei bufSize, void* pixels);
+			static void extGlGetTextureImage(GLuint texture, GLenum target, GLint level, GLenum format, GLenum type, GLsizei bufSizeHint, void* pixels);
+			static void extGlGetCompressedTextureImage(GLuint texture, GLenum target, GLint level, GLsizei bufSizeHint, void* pixels);
+			static void extGlTextureSubImage1D(GLuint texture, GLenum target, GLint level, GLint xoffset, GLsizei width, GLenum format, GLenum type, const void* pixels);
+			static void extGlTextureSubImage2D(GLuint texture, GLenum target, GLint level, GLint xoffset, GLint yoffset, GLsizei width, GLsizei height, GLenum format, GLenum type, const void* pixels);
+			static void extGlTextureSubImage3D(GLuint texture, GLenum target, GLint level, GLint xoffset, GLint yoffset, GLint zoffset, GLsizei width, GLsizei height, GLsizei depth, GLenum format, GLenum type, const void* pixels);
+			static void extGlCompressedTextureSubImage1D(GLuint texture, GLenum target, GLint level, GLint xoffset, GLsizei width, GLenum format, GLsizei imageSize, const void* data);
+			static void extGlCompressedTextureSubImage2D(GLuint texture, GLenum target, GLint level, GLint xoffset, GLint yoffset, GLsizei width, GLsizei height, GLenum format, GLsizei imageSize, const void* data);
+			static void extGlCompressedTextureSubImage3D(GLuint texture, GLenum target, GLint level, GLint xoffset, GLint yoffset, GLint zoffset, GLsizei width, GLsizei height, GLsizei depth, GLenum format, GLsizei imageSize, const void* data);
+			static void extGlGenerateTextureMipmap(GLuint texture, GLenum target);
+			static void extGlTextureParameterIuiv(GLuint texture, GLenum target, GLenum pname, const GLuint* params);
+			static void extGlCreateSamplers(GLsizei n, GLuint* samplers);
+			static void extGlBindSamplers(const GLuint& first, const GLsizei& count, const GLuint* samplers);
+			static void extGlBindImageTextures(GLuint first, GLsizei count, const GLuint* textures, const GLenum* formats);
+			static GLuint64 extGlGetTextureHandle(GLuint texture);
+			static GLuint64 extGlGetTextureSamplerHandle(GLuint texture, GLuint sampler);
+			static void extGlMakeTextureHandleResident(GLuint64 handle);
+			static void extGlMakeTextureHandleNonResident(GLuint64 handle);
+			static GLuint64 extGlGetImageHandle(GLuint texture, GLint level, GLboolean layered, GLint layer, GLenum format);
+			static void extGlMakeImageHandleResident(GLuint64 handle, GLenum access);
+			static void extGlMakeImageHandleNonResident(GLuint64 handle);
+			static GLboolean extGlIsTextureHandleResident(GLuint64 handle);
+			static GLboolean extGlIsImageHandleResident(GLuint64 handle);
+			static void extGlCreateFramebuffers(GLsizei n, GLuint* framebuffers);
+			static GLenum extGlCheckNamedFramebufferStatus(GLuint framebuffer, GLenum target);
+			static void extGlNamedFramebufferTexture(GLuint framebuffer, GLenum attachment, GLuint texture, GLint level);
+			static void extGlNamedFramebufferTextureLayer(GLuint framebuffer, GLenum attachment, GLuint texture, GLenum textureType, GLint level, GLint layer);
+			static void extGlBlitNamedFramebuffer(GLuint readFramebuffer, GLuint drawFramebuffer, GLint srcX0, GLint srcY0, GLint srcX1, GLint srcY1, GLint dstX0, GLint dstY0, GLint dstX1, GLint dstY1, GLbitfield mask, GLenum filter);
+			static void extGlActiveStencilFace(GLenum face);
+			static void extGlNamedFramebufferReadBuffer(GLuint framebuffer, GLenum mode);
+			static void extGlNamedFramebufferDrawBuffer(GLuint framebuffer, GLenum buf);
+			static void extGlNamedFramebufferDrawBuffers(GLuint framebuffer, GLsizei n, const GLenum* bufs);
+			static void extGlClearNamedFramebufferiv(GLuint framebuffer, GLenum buffer, GLint drawbuffer, const GLint* value);
+			static void extGlClearNamedFramebufferuiv(GLuint framebuffer, GLenum buffer, GLint drawbuffer, const GLuint* value);
+			static void extGlClearNamedFramebufferfv(GLuint framebuffer, GLenum buffer, GLint drawbuffer, const GLfloat* value);
```
revert the change to `static`

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -562,7 +582,7 @@ namespace irr {
 												GL_TEXTURE_CUBE_MAP_ARRAY,GL_TEXTURE_2D_MULTISAMPLE,GL_TEXTURE_2D_MULTISAMPLE_ARRAY }; // GL 4.x
 
-			if (COpenGLExtensionHandler::Version >= 440 || COpenGLExtensionHandler::FeatureAvailable[COpenGLExtensionHandler::EOpenGLFeatures::IRR_ARB_multi_bind])
+			if (COpenGLDriver::Version >= 440 || COpenGLDriver::FeatureAvailable[COpenGLDriver::EOpenGLFeatures::IRR_ARB_multi_bind])
```
if you create that `COpenGLFeatureMap` class and move all the Version + Feature variables there, then you will no longer have the circular dependency, because here you'd be using `COpenGLFeatureMap::` wherever you're currently trying to use `COpenGLDriver::` and you wouldn't need to know anything about `COpenGLDriver`

devshgraphicsprogramming commented on source/Irrlicht/COpenGLExtensionHandler.h:
```diff
@@ -1,4557 +1,4052 @@
+//		}
+//
+//		inline void COpenGLExtensionHandler::extGlGetTextureImage(GLuint texture, GLenum target, GLint level, GLenum format, GLenum type, GLsizei bufSizeHint, void* pixels)
+//		
```
just remove the file

devshgraphicsprogramming commented on source/Irrlicht/COpenGLExtensionHandler.h:
```diff
@@ -1,4557 +1,4052 @@
+//			"GL_NVX_multiview_per_view_attributes"
+//		};
+//		//extra extension name that is reported as supported when irrbaw app is running in renderdoc
+//		_IRR_STATIC_INLINE_CONSTEXPR const char* RUNNING_IN_RENDERDOC_EXTENSION_NAME = "GL_IRR_RUNNING_IN_RENDERDOC";
```
move this to `COpenGLFeatureMap`

Hazardu committed:
COpenGLFeatureMap

Hazardu committed:
Passing Function Table ptr to SAuxContext

devshgraphicsprogramming commented on include/irr/system/DynamicLibraryFunctionPointer.h:
```diff
@@ -46,7 +46,7 @@ class DynamicLibraryFunctionPointer
 			return error(name);
 		}
-
+		/*
```
what;s the reason for this change?

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -222,15 +222,14 @@ namespace irr
 COpenGLDriver::COpenGLDriver(const irr::SIrrlichtCreationParameters& params,
 		io::IFileSystem* io, CIrrDeviceWin32* device, const asset::IGLSLCompiler* glslcomp)
-: CNullDriver(device, io, params), //COpenGLFunctionTable(),   // What to do here?
+: CNullDriver(device, io, params), COpenGLFunctionTable(),
```
why does `COpenGLDriver` inherit from `COpenGLFunctionTable` ....

Function table is supposed to be used by composition, not inheritance, and its supposed to be in the `SAuxContext` only

devshgraphicsprogramming commented on src/irr/video/COpenGLFeatureMap.h:
```diff
@@ -0,0 +1,1113 @@
+	{
+
+
+static const char* const OpenGLFeatureStrings[] = {
```
move it into the `COpenGLFeatureMap` as a member, so it doesn't pollute the namespace as a global

devshgraphicsprogramming commented on src/irr/video/COpenGLFunctionTable.h:
```diff
@@ -1829,14 +1827,14 @@ namespace irr {
 		}
 
-		inline GLboolean COpenGLFunctionTable::extGlUnmapNamedBuffer(GLuint buffer)
+		GLboolean COpenGLFunctionTable::extGlUnmapNamedBuffer(GLuint buffer)
```
KEEP THE `inline` everywhere!

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.h:
```diff
@@ -204,513 +204,7 @@ template<> struct pipeline_for_bindpoint<EPBP_GRAPHICS> { using type = COpenGLRe
-};
-
-class COpenGLDriver final : public CNullDriver
+class COpenGLDriver final : public CNullDriver, public COpenGLFeatureMap, public COpenGLFunctionTable
```
stop inheriting from function table!

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.h:
```diff
@@ -1987,6 +899,8 @@ class COpenGLDriver final : public CNullDriver
         //private:
             std::thread::id threadId;
             uint8_t ID; //index in array of contexts, just to be easier in use
+			COpenGLFunctionTable* FunctionTable;	//Function Table Object
```
no need for it to be a pointer... just let it be a member

Hazardu committed:
enable_if in  DynamicLibraryFunctionPointer.h

Hazardu committed:
Changed function table to be used from SAuxContext, problem with getting context ptr in COpenGLDriver.h line 475

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -635,7 +635,7 @@ bool COpenGLDriver::initDriver(CIrrDeviceWin32* device)
 	genericDriverInit(device->getAssetManager());
 
-	COpenGLFunctionTable::extGlSwapInterval(Params.Vsync ? 1 : 0);
+    getThreadContext_helper(true)->functionTable.extGlSwapInterval(Params.Vsync ? 1 : 0);
```
did you already lock the context mutex? because thats the only time you should call that function with `true`

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.h:
```diff
@@ -470,7 +471,8 @@ class COpenGLDriver final : public CNullDriver, public COpenGLFeatureMap, public
 				{
 					GLint res = GL_FALSE;
-					glGeneral.pglGetInternalformativ(GL_TEXTURE_2D, getSizedOpenGLFormatFromOurFormat(_fmt), GL_COLOR_RENDERABLE, sizeof(res), &res);
+					//this is troublesome, how to obtain context reference here?
+					//functionTable.glGeneral.pglGetInternalformativ(GL_TEXTURE_2D, getSizedOpenGLFormatFromOurFormat(_fmt), GL_COLOR_RENDERABLE, sizeof(res), &res);
```
its not a static function, so it should be fine, the only problem is the `const`.... @Crisspl any advice?

Maybe we could make the `operator()` on the function pointer wrapper class a `const` operator?
